File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        @cache
        def recur(n, sub_sum):
            if sub_sum == 0:
                return True

            if n == 0 or sub_sum < 0:
                return False

            result = recur(n - 1, sub_sum - nums[n - 1]) or recur(n - 1, sub_sum)

            return result

        n = len(nums)

        total = sum(nums)
        if total % 2 != 0:
            return False

        subset_sum = total // 2
        n = len(nums)
        return recur(n - 1, subset_sum)

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.ts

function canPartition(nums: number[]): boolean {
    const recur = (n: number, subSum: number) => {
        if (subSum === 0) return true;

        if (n === 0 || subSum < 0) return false;

        if (memo.has(`${n}-${subSum}`)) return memo.get(`${n}-${subSum}`)
        const result = recur(n - 1, subSum - nums[n - 1]) || recur(n - 1, subSum);
        // console.log(memo, result)
        memo.set(`${n}-${subSum}`, result);
        return result;
    }
    
    const n = nums.length;
    const total = nums.reduce((acc, num) => acc + num, 0);
    if (total % 2 !== 0) return false;
    const subSum = Math.floor(total / 2);
    const memo = new Map();
    return recur(n, subSum);
};